sys/usb/usbus/msc/scsi.c

```diff
@@ -0,0 +1,386 @@
+                                           handler_ctrl);
+    msc_cbw_rw10_pkt_t *pkt = (msc_cbw_rw10_pkt_t *)cbw->cb;
+    mtd_dev_t *mtd = msc->lun_dev[lun].mtd;
+    uint32_t mtd_sector_size = mtd->page_size * mtd->pages_per_sector;
+
+    /* Convert big endian data from SCSI to little endian */
+    /* Get first block number to read from */
+    if (mtd_sector_size < 512) {
+        msc->block = byteorder_ntohl(pkt->blk_addr) * (512 / mtd_sector_size);
+    }
+    else if (mtd_sector_size > 4096) {
+        /* TODO: add support */
+        assert(0);
+    }
+    else {
+        msc->block = byteorder_ntohl(pkt->blk_addr);
+    }
+
+    /* Get number of blocks to transfer */
+    msc->block_nb = byteorder_ntohs(pkt->xfer_len);
+    msc->cmd.len = cbw->data_len;
+    msc->state = DATA_TRANSFER_OUT;
+}
+
+static void _scsi_read10(usbus_handler_t *handler, msc_cbw_buf_t *cbw, uint8_t lun)
+{
+    usbus_msc_device_t *msc = container_of(handler, usbus_msc_device_t,
+                                           handler_ctrl);
+    msc_cbw_rw10_pkt_t *pkt = (msc_cbw_rw10_pkt_t *)cbw->cb;
+    mtd_dev_t *mtd = msc->lun_dev[lun].mtd;
+    uint32_t mtd_sector_size = mtd->page_size * mtd->pages_per_sector;
+    /* Convert big endian data from SCSI to little endian */
+    /* Get first block number to read from */
+    if (mtd_sector_size < 512) {
+        /* If MTD sector size is less than 512 (e.g for EEPROM),
+         * apply some logic and let host believe this is a 512
+         * sector size device and let RIOT write several sectors in
+         * a row to reach 512 bytes so everyone is happy */
+        msc->block = byteorder_ntohl(pkt->blk_addr) * (512 / mtd_sector_size);
+    }
+    else if (mtd_sector_size > 4096) {
+        /* If MTD sector size is above 4096 bytes (e.g some MCU internal flash)
+         * limits the reported block size to 4096 bytes and add some logic to
+         * write less than a sector (using read-modify-write method). This will
+         * reduce the overall performance !*/
+        msc->block = byteorder_ntohl(pkt->blk_addr) / (mtd_sector_size / 4096);
+    }
+    else {
+        msc->block = byteorder_ntohl(pkt->blk_addr);
+    }
+
+    /* Get number of blocks to transfer */
+    msc->block_nb = byteorder_ntohs(pkt->xfer_len);
+    msc->cmd.len = cbw->data_len;
+    msc->state = DATA_TRANSFER_IN;
+
+    if ((cbw->flags & USB_MSC_CBW_FLAG_IN) != 0) {
+        _rx_ready(msc);
+    }
+}
+
+static void _scsi_inquiry(usbus_handler_t *handler)
+{
+    usbus_msc_device_t *msc = container_of(handler, usbus_msc_device_t,
+                                           handler_ctrl);
+    msc_inquiry_pkt_t *pkt = (msc_inquiry_pkt_t *)msc->in_buf;
+    size_t len = sizeof(msc_inquiry_pkt_t);
+
+    /* Clear the exact amount of bytes needed for this transfer */
+    memset(pkt, 0, len);
+
+    /* prepare pkt response */
+    pkt->type = 0;
+    pkt->removable = 1;
+    pkt->version = 0x01;
+    pkt->length = len - 4;
```

Comment:
Why `- 4`?

Reply:
This is the additional packet length minus the "default" data of this packet.
a SCSI inquiry response will always include the 4 first bytes, but you may have to pass additional information. Thus, you pass the number additional bytes through this field.

I'll add a comment.